pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -657,7 +657,10 @@ def _validate_monitor_key(self, trainer: 'pl.Trainer') -> None:
                 f" {list(metrics.keys())}. "
                 f"HINT: Did you call self.log('{self.monitor}', value) in the LightningModule?"
             )
-            raise MisconfigurationException(m)
+            if not trainer.fit_loop.val_loop._has_run:
+                warning_cache.warn(m)
+            else:
+                raise MisconfigurationException(m)
```

Comment:
@tchaton could you explain this check further? why does the val loop running need to be checked here for a warning vs an exception? relying on the val loop's property is confusing on first read since we support checkpointing during the training loop (every n batches)

Reply:
Hey @ananthsub, yes definitely !

ModelCheckpoint might rely on "val_loss" to checkpoint which is generated by validation_step.

We had flicky test relying on sanity_checking to assert the monitor value actually existed.

However, as the monitor metric doesn't exist until val_loop has been run, the ModelCheckpoint should just warn and skip checkpointing if the key is missing.

When val_loop has actually been run, then it should raise an exception if the metric key is missing (due to a typo or missing for step). 

If you rely on the metric generated by training_step, then the key will be present independently of  trainer.fit_loop.val_loop._has_run value and no warning or exception will be triggered.

Best,
T.C

Reply:
> We had flicky test relying on sanity_checking to assert the monitor value actually existed.

But we don't save checkpoints during sanity checking at all. why'd it reach this point? is it possible for us to remove the `_has_run` property and its check here?
- having separate checks for train loop vs val loop is harder to maintain
- we can check if the trainer is sanity checking via the trainer property inside the callback
- `_has_run` is another form of progress tracking which sits outside the core tracking dataclasses, which we want to converge to

Reply:
I overlooked this.
I too think that the `_has_run` field here is a hack that we want to solve.
The `ModelCheckpoint` already tests against `trainer.sanity_checking`.  If there is still a reason for checking if the loop was run, we should rely on progress tracking.

Reply:
Yes, I agree. Once progress tracking has been added, this should be updated.